pkg/streamingpromql/types/limiting_pool.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package types
+
+import (
+	"unsafe"
+
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+const (
+	maxExpectedPointsPerSeries  = 100_000 // There's not too much science behind this number: 100000 points allows for a point per minute for just under 70 days.
+	pointsPerSeriesBucketFactor = 2.0
+
+	// Treat a native histogram sample as equivalent to this many float samples when considering max in-memory bytes limit.
+	// Keep in mind that float sample = timestamp + float value, so 5x this is equivalent to five timestamps and five floats.
+	nativeHistogramSampleSizeFactor = 5
+
+	FPointSize           = uint64(unsafe.Sizeof(promql.FPoint{}))
+	HPointSize           = uint64(FPointSize * nativeHistogramSampleSizeFactor)
+	VectorSampleSize     = uint64(unsafe.Sizeof(promql.Sample{})) // This assumes each sample is a float sample, not a histogram.
+	Float64Size          = uint64(unsafe.Sizeof(float64(0)))
+	BoolSize             = uint64(unsafe.Sizeof(false))
+	HistogramPointerSize = uint64(unsafe.Sizeof((*histogram.FloatHistogram)(nil)))
+)
+
+var (
+	FPointSlicePool = NewLimitingBucketedPool(
```

Comment:
Why maintain these globals?

We could still have a Pool separate from the MemoryConsumptionTracker that we pass around as needed.

Reply:
What advantage do we gain from passing around a separate `Pool`?

Reply:
Not having globals :-p.

Globals have a few anti-patterns to them that we generally like to avoid.

Reply:
In general, I agree with you. In this particular case, we need to have a global somewhere for the pool, otherwise we can't share the pool between queries, and using a global variable like this is fairly common in Go for this purpose.

Reply:
If it's a common pattern that's fine by me :+1: